inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/non_max_suppression_gpu_ref.cl

```diff
@@ -0,0 +1,571 @@
+        outBoxes[count] = binfo;
+        ++count;
+    }
+
+    return count;
+}
+
+inline void FUNC(initOutputBoxList)(__global BOX_INFO *outBoxes, int boxNum, const __global INPUT1_TYPE *scores, __global OUTPUT_TYPE *output)
+{
+    for (int i = 0; i < boxNum; ++i) {
+        const int outputId = i * 3;
+        outBoxes[i].batchId = output[outputId + 0];
+        outBoxes[i].classId = output[outputId + 1];
+        outBoxes[i].boxId = output[outputId + 2];
+        outBoxes[i].score = scores[INPUT1_GET_INDEX(outBoxes[i].batchId, outBoxes[i].classId, outBoxes[i].boxId, 0)];
+    }
+}
+
+inline void FUNC(swap)(__global BOX_INFO* a, __global BOX_INFO* b)
+{
+    BOX_INFO temp = *a;
+    *a = *b;
+    *b = temp;
+}
+
+inline void FUNC(sortOutputBoxList)(__global BOX_INFO *outSortedBoxes, int boxNum)
+{
+    for (int i = 0; i < boxNum - 1; ++i) {
+        bool swapped = false;
+        for (int j = 0; j < boxNum - i - 1; ++j) {
+            if ((outSortedBoxes[j].score < outSortedBoxes[j+1].score) ||
+                (outSortedBoxes[j].score == outSortedBoxes[j+1].score && outSortedBoxes[j].batchId > outSortedBoxes[j+1].batchId) ||
+                (outSortedBoxes[j].score == outSortedBoxes[j+1].score && outSortedBoxes[j].batchId == outSortedBoxes[j+1].batchId &&
+                 outSortedBoxes[j].classId > outSortedBoxes[j+1].classId) ||
+                (outSortedBoxes[j].score == outSortedBoxes[j+1].score && outSortedBoxes[j].batchId == outSortedBoxes[j+1].batchId &&
+                 outSortedBoxes[j].classId == outSortedBoxes[j+1].classId && outSortedBoxes[j].boxId > outSortedBoxes[j+1].boxId)) {
+                FUNC_CALL(swap)(&outSortedBoxes[j], &outSortedBoxes[j+1]);
+                swapped = true;
+            }
+        }
+
+        // IF no two elements were swapped by inner loop, then break
+        if (swapped == false)
+            break;
+    }
+}
+
+// KERNEL_ARGs
+//
+// boxes
+//  - shape: {num_batches, num_boxes, 4}
+// scores
+//  - shape: {num_batches, num_classes, num_boxes}
+// buffer0 (intermediate buffer)
+//  - size: batch_num * class_num * boxes_num * sizeof(SBOX_INFO)
+//  - desc: filtered and sorted SBOX_INFO list
+// buffer1 (intermediate buffer)
+//  - size: batch_num * class_num * boxes_num * sizeof(BOX_INFO)
+//  - desc: selected SBOX_INFO list by iou calucation
+// buffer2 (intermediate buffer)
+//  - size: batch_num * class_num * boxes_num * sizeof(BOX_INFO)
+//  - desc: selected SBOX_INFO list by iou calucation
+// buffer3 (intermediate buffer)
+//  - size: batch_num * class_num * 4
+//  - desc: sorted box num for batch*class
+#ifdef IS_STAGE_0
+KERNEL (non_max_suppression_ref_stage_0)(
+    const __global INPUT1_TYPE *scores
+    , __global uchar *buffer0
+    , __global int *buffer3
+    #ifdef SCORE_THRESHOLD_TYPE
+    , const __global SCORE_THRESHOLD_TYPE *score_threshold
+    #endif
+    )
+{
+    const short batchId = get_global_id(0);
+    const ushort classId = get_global_id(1);
```

Comment:
@vladimir-paramuzov  short range is up to 32767. Do you think batch or classId may exceed the number?

Reply:
@isanghao We can use ushort, but we definitely need a check on host side that shape doesn't exceed data type range to avoid UB for such cases. Or (even better) parameterize this type based on input shape.

Reply:
I haven't found the check for batch/feature values which will prevent usage of this impl if range doesn't fit short/ushort precision. It it added?